### ISSUE TITLE
Tflite does not seem to support LSTM.

### ISSUE THREAD

**Issue by Arsenal591** [tensorflow/tensorflow#21377]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: Yes
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: CentOS Linux release 7.3.1611
- **Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device**: N/A
- **TensorFlow installed from (source or binary)**: binary
- **TensorFlow version (use command below)**: 1.9.0
- **Python version**: 3.6.6
- **Bazel version (if compiling from source)**: N/A
- **GCC/Compiler version (if compiling from source)**: N/A
- **CUDA/cuDNN version**: cuda=8.0.61 cudnn=7.0.5
- **GPU model and memory**: N/A
- **Exact command to reproduce**: See below.

### Source code / logs
See below.

### Describe the problem
I'm trying to convert a `.pb` graph into tflite format, which containes a bi-directional LSTM. However, whenever I run the command:
```
toco --graph_def_file=opt_frozen.pb \
--output_file= output.tflite \
--output_format=TFLITE \
--inference_type=FLOAT \
--inference_input_type=FLOAT \
--input_arrays=[input_tensor_names] \
--output_arrays=[output_tensor_names]
```
It failed with following error messages:
```
2018-08-03 14:35:41.259412: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1366] Converting unsupported operation: Enter
2018-08-03 14:35:41.259490: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1366] Converting unsupported operation: Enter
2018-08-03 14:35:41.271357: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1366] Converting unsupported operation: Enter
2018-08-03 14:35:41.271401: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1366] Converting unsupported operation: Enter
2018-08-03 14:35:41.271439: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1366] Converting unsupported operation: Enter
2018-08-03 14:35:41.271454: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1366] Converting unsupported operation: Enter
2018-08-03 14:35:41.279573: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1366] Converting unsupported operation: Enter
2018-08-03 14:35:41.279626: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1366] Converting unsupported operation: Enter
2018-08-03 14:35:41.279989: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1366] Converting unsupported operation: Enter
2018-08-03 14:35:41.280020: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1366] Converting unsupported operation: Enter
2018-08-03 14:35:41.280043: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1366] Converting unsupported operation: Enter
2018-08-03 14:35:41.280087: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1366] Converting unsupported operation: TensorArrayV3
2018-08-03 14:35:41.280107: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1366] Converting unsupported operation: TensorArrayV3
2018-08-03 14:35:41.280123: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1366] Converting unsupported operation: Enter
2018-08-03 14:35:41.280141: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1366] Converting unsupported operation: Enter
2018-08-03 14:35:41.280163: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1366] Converting unsupported operation: LogicalAnd
2018-08-03 14:35:41.280175: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1366] Converting unsupported operation: LoopCond
2018-08-03 14:35:41.280224: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1366] Converting unsupported operation: Enter
2018-08-03 14:35:41.280266: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1366] Converting unsupported operation: Enter
2018-08-03 14:35:41.280303: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1366] Converting unsupported operation: Exit
2018-08-03 14:35:41.280317: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1366] Converting unsupported operation: TensorArraySizeV3
2018-08-03 14:35:41.280340: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1366] Converting unsupported operation: TensorArrayGatherV3
2018-08-03 14:35:41.280362: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1366] Converting unsupported operation: ReverseV2
2018-08-03 14:35:41.280385: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1366] Converting unsupported operation: Enter
2018-08-03 14:35:41.280412: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1366] Converting unsupported operation: Enter
2018-08-03 14:35:41.280433: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1366] Converting unsupported operation: Enter
2018-08-03 14:35:41.280469: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1366] Converting unsupported operation: TensorArrayV3
2018-08-03 14:35:41.280487: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1366] Converting unsupported operation: TensorArrayV3
2018-08-03 14:35:41.280502: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1366] Converting unsupported operation: Enter
2018-08-03 14:35:41.280519: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1366] Converting unsupported operation: Enter
2018-08-03 14:35:41.280535: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1366] Converting unsupported operation: LogicalAnd
2018-08-03 14:35:41.280546: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1366] Converting unsupported operation: LoopCond
2018-08-03 14:35:41.280588: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1366] Converting unsupported operation: Enter
2018-08-03 14:35:41.280626: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1366] Converting unsupported operation: Enter
2018-08-03 14:35:41.280656: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1366] Converting unsupported operation: Exit
2018-08-03 14:35:41.280671: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1366] Converting unsupported operation: TensorArraySizeV3
2018-08-03 14:35:41.280693: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1366] Converting unsupported operation: TensorArrayGatherV3
2018-08-03 14:35:41.280965: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1366] Converting unsupported operation: TensorArrayScatterV3
2018-08-03 14:35:41.280982: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1366] Converting unsupported operation: Enter
2018-08-03 14:35:41.280997: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1366] Converting unsupported operation: TensorArrayReadV3
2018-08-03 14:35:41.281074: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1366] Converting unsupported operation: TensorArrayWriteV3
2018-08-03 14:35:41.281102: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1366] Converting unsupported operation: ReverseV2
2018-08-03 14:35:41.281123: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1366] Converting unsupported operation: TensorArrayScatterV3
2018-08-03 14:35:41.281138: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1366] Converting unsupported operation: Enter
2018-08-03 14:35:41.281157: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1366] Converting unsupported operation: TensorArrayReadV3
2018-08-03 14:35:41.281228: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1366] Converting unsupported operation: TensorArrayWriteV3
```
And I assume that LSTM is not currently supported in tflite.
So is that the case? And if it's true, is there any way that I can work around this problem?

**Comment by aselle** [comment]
You can generate a LSTM cell indidivually and manually loop back state (i.e. don't use static_rnn or dynamic_rnn). It will convert the lstm as individual operators. For more help, please provide a simple example that you are trying to run.

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting response` label was assigned. Is this still an issue?

**Comment by tofulawrence** [comment]
Automatically closing due to lack of recent activity. Please update the issue when new information becomes available, and we will reopen the issue. Thanks!

**Comment by binbinsh** [comment]
> You can generate a LSTM cell indidivually and manually loop back state (i.e. don't use static_rnn or dynamic_rnn). It will convert the lstm as individual operators. For more help, please provide a simple example that you are trying to run.

Hi @aselle, I am trying to make my TFLiteLSTMCell work with TFLite interpreter, I trained the model with dynamic_rnn and restore it as static_rnn when using TFLiteConverter. I was successfully convert the model into .tflite file and run successfully, but I have to fix the sequence length when running TFLiteConverter with static_rnn.

As you mentioned may be I can generate a LSTM cell individually and manually loop back state. How can I achieve this? could you give me an example on how to do this?

I define the LSTM cell like this:
```
lstm_cell1 = TFLiteLSTMCell(num_units=128, initializer=tf.orthogonal_initializer())
lstm_cell2 = TFLiteLSTMCell(num_units=128, initializer=tf.orthogonal_initializer())
multi_lstm_cell = tf.contrib.rnn.MultiRNNCell([lstm_cell1, lstm_cell2])

# when training
outputs, state = tf.nn.dynamic_rnn(cell=multi_lstm_cell,
 input=xxx, sequence_length=yyy, initial_state=zzz)

# when converting
outputs, state = tf.nn.static_rnn(...)
```

**Comment by robbiemu** [comment]
@aselle  I have a simple example to volunteer :) I wish you would take mine as an example! Full source is https://colab.research.google.com/drive/1sB8vv3TM5Sp45TicJ8PxEvpUAX42XqPY#scrollTo=t5AMh3Ui115y&uniqifier=1
 and I can gladly provide sample data as well.

Here's the part specific to consrtructing the model:

```python
from keras.models import Sequential
from keras.layers import Dense, LSTM, Dropout

regressor = Sequential()
regressor.add(LSTM(units=nh, return_sequences=True, 
                   input_shape=(timesteps, dim)))

layers = (len(endog) + 1) if len(endog) > 1 else 2 # 4
print('layers', layers)
for i in range(1, layers):
  if i < layers - 1:
    cell = LSTM(units=nh, return_sequences=True)
  else:
    cell = LSTM(units=nh)

  regressor.add(cell)
  
regressor.add(Dropout(dropout)) # 0.2
regressor.add(Dense(units=dim))
              
regressor.compile(optimizer='adam', loss='mean_squared_error')
```
